components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/SyncableLoginsStorage.kt

```diff
@@ -166,129 +154,98 @@ class SyncableLoginsStorage(
+        store.addOrUpdate(entry.toAS(), key).toAC()
     }
 
     override fun registerWithSyncManager() {
-        conn.getStorage().registerWithSyncManager()
+        store.registerWithSyncManager()
     }
 
     override fun getHandle(): Long {
         throw NotImplementedError("Use registerWithSyncManager instead")
     }
 
     /**
+     * @throws [CryptoException] invalid encryption key
      * @throws [LoginsStorageException] If DB isn't empty during an import; also, on unexpected errors
      * (IO failure, rust panics, etc).
      */
-    @Throws(LoginsStorageException::class)
+    @Throws(CryptoException::class, LoginsStorageException::class)
     override suspend fun importLoginsAsync(logins: List<Login>): JSONObject = withContext(coroutineContext) {
-        JSONObject(conn.getStorage().importLogins(logins.map { it.toDatabaseLogin() }))
+        JSONObject(store.importMultiple(logins.map { it.toAS() }, key))
     }
 
     /**
-     * @throws [InvalidRecordException] On both expected errors (malformed [login], [login]
-     * already exists in store, etc. See [InvalidRecordException.reason] for details) and
-     * unexpected errors (IO failure, rust panics, etc)
+     * @throws [LoginsStorageException] On unexpected errors (IO failure, rust panics, etc)
      */
-    @Throws(InvalidRecordException::class)
-    override suspend fun ensureValid(login: Login) = withContext(coroutineContext) {
-        conn.getStorage().ensureValid(login.toDatabaseLogin())
+    @Throws(LoginsStorageException::class)
+    override suspend fun getByBaseDomain(origin: String): List<EncryptedLogin> = withContext(coroutineContext) {
+        store.getByBaseDomain(origin).map { it.toAC() }
     }
 
     /**
+     * @throws [CryptoException] invalid encryption key
      * @throws [LoginsStorageException] On unexpected errors (IO failure, rust panics, etc)
      */
     @Throws(LoginsStorageException::class)
-    override suspend fun getByBaseDomain(origin: String): List<Login> = withContext(coroutineContext) {
-        conn.getStorage().getByBaseDomain(origin).map { it.toLogin() }
+    override fun findLoginToUpdate(entry: LoginEntry, logins: List<Login>): Login? {
+        return mozilla.appservices.logins.findLoginToUpdate(entry.toAS(), logins.map { it.toAS() })?.toAC()
     }
 
     /**
-     * @throws [LoginsStorageException] On unexpected errors (IO failure, rust panics, etc)
+     * @throws [CryptoException] invalid encryption key
      */
-    @Throws(LoginsStorageException::class)
-    override suspend fun getPotentialDupesIgnoringUsername(login: Login): List<Login> =
-        withContext(coroutineContext) {
-            conn.getStorage().potentialDupesIgnoringUsername(login.toDatabaseLogin())
-                .map { it.toLogin() }
+    override suspend fun decryptLogins(logins: List<EncryptedLogin>): List<Login> = withContext(coroutineContext) {
+        logins.map {
+            mozilla.appservices.logins.decryptLogin(it.toAS(), key).toAC()
```

Comment:
See my PR-level comment about the overall shape of the crypto stuff.

Reply:
In general, the logins code is more complicated and this made us rethink how encryption should work.  I think the long-term goal might be to move credit-cards to being more like logins.

A good example is the de-duping code.  That needs to match the login data that's being saved against a list of existing logins.  It's nicer for that to be a list of decrypted logins rather than having a list of encrypted logins and decrypting the usernames for each pass.

Another one is autofill.  You need decrypt the username to display the list, then once the user selects an item you need to decrypt again.  It seemed to us that it was better do do a single  `List<EncryptedLogin> -> List<Login>` conversion, than to decrypt multiple times, or to have a `Login` class that might or might not have decrypted data

You bring up a really good point about the data marshalling though.  We should change the logins API so that only the username/password is sent across the FFI.  The code in a-s is setup to do exactly that.  I think the `concept.storage.LoginsStorage` interface could stay the same though.